assets/src/edit-story/components/form/text.js

```diff
@@ -106,79 +106,91 @@ Clear.propTypes = {
-  placeholder,
-  ...rest
-}) {
-  const { suffix } = rest;
-  const inputRef = useRef();
-  const isMultiple = value === MULTIPLE_VALUE;
-  value = isMultiple ? '' : value;
-  placeholder = isMultiple ? MULTIPLE_DISPLAY_VALUE : placeholder;
-
-  const onClear = useCallback(() => {
-    onChange('');
-    if (onBlur) {
-      onBlur({ onClear: true });
-    }
-    // Return focus to text input as otherwise focus will revert to current selection
-    inputRef.current?.focus();
-  }, [onChange, onBlur]);
-
-  return (
-    <Container
-      className={`${className}`}
-      flexBasis={flexBasis}
-      disabled={disabled}
-      suffix={suffix}
-    >
-      {/* type="text" is default but added here due to an a11y-related bug. See https://github.com/A11yance/aria-query/pull/42 */}
-      <StyledInput
-        type="text"
-        ref={inputRef}
-        placeholder={placeholder}
-        label={label}
-        value={value}
+const TextInput = forwardRef(
+  (
+    {
+      className,
+      onBlur,
+      onChange,
+      onKeyDown,
+      label,
+      value,
+      flexBasis,
+      disabled,
+      clear,
+      clearIcon,
+      showClearIconBackground,
+      placeholder,
+      ...rest
+    },
+    ref
+  ) => {
+    const { suffix } = rest;
+    const inputRef = useRef();
+    const isMultiple = value === MULTIPLE_VALUE;
+    value = isMultiple ? '' : value;
+    placeholder = isMultiple ? MULTIPLE_DISPLAY_VALUE : placeholder;
+
+    const onClear = useCallback(() => {
+      onChange('');
+      if (onBlur) {
+        onBlur({ onClear: true });
+      }
+      // Return focus to text input as otherwise focus will revert to current selection
+      inputRef.current?.focus();
+    }, [onChange, onBlur]);
+
+    return (
+      <Container
+        className={`${className}`}
+        flexBasis={flexBasis}
         disabled={disabled}
-        {...rest}
-        onChange={(evt) => onChange(evt.target.value, evt)}
-        onKeyDown={onKeyDown}
-        onBlur={(evt) => {
-          if (evt.target.form) {
-            evt.target.form.dispatchEvent(
-              new window.Event('submit', { cancelable: true })
-            );
-          }
-          if (onBlur) {
-            onBlur(evt);
-          }
-        }}
-      />
-      {suffix}
-      {Boolean(value) && clear && (
-        <Clear
-          onClear={onClear}
-          showClearIconBackground={showClearIconBackground}
-        >
-          {clearIcon ?? (
-            <CloseIcon aria-label={__('Clear input', 'web-stories')} />
-          )}
-        </Clear>
-      )}
-    </Container>
-  );
-}
+        suffix={suffix}
+      >
+        {/* type="text" is default but added here due to an a11y-related bug. See https://github.com/A11yance/aria-query/pull/42 */}
+        <StyledInput
+          type="text"
+          ref={(node) => {
+            if (ref !== null) {
+              ref.current = node;
+            }
+            inputRef.current = node;
+          }}
```

Comment:
I'm a little confused about this. Tell me if I'm misunderstanding:

Looks like we're either setting the `ref` if it's passed in through `forwardRef`, otherwise we're setting the `inputRef` that's initialized in the `TextInput`. Is that correct?


Does `ref={ref || inputRef}` not work the same as the above?

Reply:
Doesn't `ref || inputRef` just make it so you're only assigning one `ref.current`? In the case of forwarding a ref, only the forwarded ref would have `ref.current` and `inputRef.current` would never be assigned. I wanted to retain the input's self control with `inputRef` but also be able to pass it a ref. I can still clean it up to have the forwarded ref just reference the `inputRef` above the return rather than in the prop here, though.

Reply:
hmm yeah I guess I was thinking that if we don't have `ref` passed into this component then we should use `inputRef` as a fallback. But if we need both to be the same ref then this makes sense :thumb